modin/core/dataframe/pandas/dataframe/utils.py

```diff
@@ -0,0 +1,246 @@
+    q: float = 0.1,
+    method: str = "linear",
+) -> "tuple[pandas.DataFrame, np.ndarray]":
+    """
+    Pick quantiles over the given partition.
+
+    This function is applied to each row-axis partition in parallel to select quantiles
+    from each. It samples using the following algorithm:
+        * If there are <= A (100) rows in the dataframe, we pick quantiles over the entire dataframe.
+        * If there are A (100) < # of rows <= (A * (1 - k))/(q - k) (1900), we pick quantiles over the first A (100) rows, plus a sample
+        of k (5%) of the remaining rows.
+        * If there are > (A * (1 - k))/(q - k) (1900) rows, we pick quantiles over a sample of q (10%) of all of the rows.
+
+    These numbers are a heuristic. They were picked such that the size of the sample
+    in scenario 2 scales well. In other words, we picked A, k, and q such that:
+        A + (len(df) - A)*k = q*len(df)
+    Where q is the proportion we sample in scenario 3, k is the proportion we sample
+    of remaining rows in scenario 2, and A is the threshold below which we just return
+    the entire dataframe, instead of sampling (scenario 1).
+    q = 0.1 and k = 0.05 were picked such that this formula holds for A = 100.
+
+    Parameters
+    ----------
+    df : pandas.Dataframe
+        The dataframe to pick quantiles from.
+    columns : list[str]
+        The columns to pick quantiles from. Only the first column in the list will be used.
+    A : int, default: 100
+        A heuristic that defines "small" dataframes.
+    k : float, default: 0.05
+        A heuristic used for sampling.
+    q : float, default: 0.1
+        A hueristic used for sampling.
+    method : str, default: linear
+        The method to use when picking quantiles.
+
+    Returns
+    -------
+    pandas.Dataframe, np.ndarray:
+        The partition that was provided as input, as well as the quantiles for the partition.
+
+    Notes
+    -----
+    quantiles are only computed over the first column of the sort.
+    """
+    quantiles = [i / (NPartitions.get() * 2) for i in range(NPartitions.get() * 2)]
+    # Heuristic for a "small" df we will compute quantiles over entirety of.
+    if len(df) <= A:
+        return df, np.quantile(df[columns[0]], quantiles, method=method)
+    # Heuristic for a "medium" df where we will include first 100 (A) rows, and sample
+    # of remaining rows when computing quantiles.
+    if len(df) <= (A * (1 - k)) / (q - k):
+        return df, np.quantile(
+            np.concatenate(
+                (
+                    df[columns[0]][:100].values,
+                    df[columns[0]][100:].sample(frac=k),
+                )
+            ),
+            quantiles,
+            method=method,
+        )
+    # Heuristic for a "large" df where we will sample 10% (q) of all rows to compute quantiles
+    # over.
+    return df, np.quantile(df[columns[0]].sample(frac=q), quantiles, method=method)
+
+
+def pick_pivots_from_quantiles_for_sort(
```

Comment:
It looks like we pick pivot points by taking the quantile of the quantiles? One problem with that approach is that it doesn't take into account the size differences among partitions at all. 

I think it would be good to document in the code why we're taking this approach, and why we think it's good enough. Is there some simple mathematical explanation for why the heuristic works?

Reply:
There wasn't really any heuristic behind this - it was just the first sampling algorithm we used. We now use a sampling algorithm derived from the TeraSort paper - should I just delete this sampling algo?